[Multilane] Adds support for multi-lane connections

agalbachicar opened the pull request:
This PR provides support to `multilane::Builder` to to create multi-lane segments in the context of #4934.

`multilane::Connections` were modified to accept new parameters that better describe Segment's requirements. Those are needed by the `multilane::Builder` to compute the asphalt extents as well as lanes and its lateral offset (lane's _r0_). New parameters are:

-  `num_lanes`: the number of lanes.
- `r0`: distance from the reference curve to first lane centerline.
- `left_shoulder` and `right_shoulder`: extra lateral space at the sides of the lanes.

`multilane::Builder` will create Lanes with the same `lane_width` width, their centerlines will be centered and centerlines distance between each other will be `lane_width` too. Aside from lanes, also new `multilane::BranchPoints` should be created to correctly start and end each lane. No change is needed on `multilane::BranchPoints` though.

YAML format changes will come in a subsequent PR once `multilane::Builder` code is done.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7171)
<!-- Reviewable:end -->

agalbachicar commented:
+@maddog-tri for feature review please.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KvYtVgO9fDCn2G_Ct35:b-5klkw7)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
First pass review "complete", PTAL.

---

Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 17 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcNP4iWnUYjAXUQCD4:-KvcNP4jheYXVOFMkBrV:b-eg1vax) (no related file):*
It would help to mention in the commit message that the YAML format changes will come in a subsequent PR.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcNqGQAaiegry2xYkr:-KvcNqGQAaiegry2xYks:b-4e7mfr) (no related file):*
Hmm... there is still something missing from this revised Builder interface.

The purpose of the Builder interface is to make it (relatively) easy to assemble a road network by chaining Connections together and having the Builder do a bunch of constraint satisfaction and math to create Lanes/Segments which are connected together smoothly.  What is missing here is a way to specify how Lanes are supposed to be aligned.  This is *implicit* in the `r0` parameters of calls to `Builder::Connect()`, but that means that the user has to do some arithmetic on scratch paper to get everything hooked up. There should be a way to have the Builder framework do that, and have the intended connectivity explicit in the user's code.

There's another snag, too.  What we have in this PR promotes the original `Lane`-based connections to `Segment`-based connections... and thus it enables a user to easily chain together Segments which have continuously connected *segment reference curves* --- because the `Connection::start()` and `end()` accessors, and the related constructor parameters, all refer to the reference curve.  But that's a overly restrictive.  We need to support chaining connections from Lane to Lane that allows the reference curves to be disconnected -- so that the segments can twist around different axes in superelevation!

I think a good example that captures this requirement is a two-lane highway which enters a curve to the left and then to the right (S-curve).  The road needs superelevation to bank into each turn --- but in each turn, the outside edge of the road should rise but the inside edge should remain at the "sea level" established by the straight/flat road entering the turns.  (The inside edge should not get lowered.)  This requires the axis of rotation (location of 'r0') to be at inside edge of each turn.

I started trying to work out a nice representation that captures these constraints, but I haven't gotten it quite figured out yet. Perhaps it's best in a follow-up PR anyhow.  Thoughts?

---

*[drake/automotive/maliput/multilane/builder.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvbn6nC_fEfOBtTzTPJ:-Kvbn6nC_fEfOBtTzTPK:bszi3e) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.cc#L26)):*
> ```cpp
>       angular_tolerance_(angular_tolerance) {
>   DRAKE_DEMAND(r_spacing >= 0.);
>   DRAKE_DEMAND(elevation_bounds.min() <= elevation_bounds.max());
> ```

This assertion already occurs within `HBounds()` itself.

---

*[drake/automotive/maliput/multilane/builder.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbvRUBwvn6DdNHFiO3:-KvbvRUBwvn6DdNHFiO4:bus6qqn) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.cc#L124)):*
> ```cpp
> // Computes the location and heading of a `lane` at given `end` creating an
> // Endpoint with that information. `zpoint` is passed here as is already
> // computed.
> ```

This use of the connection's/segment's `zpoint` here is incorrect.  It gives the `z` and `z'` values on the Segment's reference curve, but if superelevation is non-zero, then those values will be different on a Lane offset laterally.

---

*[drake/automotive/maliput/multilane/builder.h, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbgmUW8xq8mg8iVX9Q:-KvbgmUW8xq8mg8iVX9R:b-cy0bsq) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L51)):*
> ```objc
> /// explains the naming convention:
> ///
> ///  - Junctions: "j:" + Connection::id() or "j:" + Group::id()
> ```

Suggestion for clarity: 
```
  - Junctions: "j:" + Group::id(), or "j" + Connection::id() for an ungrouped Connection.
```
and then remove the `Note: when Connections are ungrouped...` comment below, too.

---

*[drake/automotive/maliput/multilane/builder.h, line 202 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbhMpBhH3a4x8ua1Pf:-KvbhMpBhH3a4x8ua1Pg:bnjnfnw) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L202)):*
> ```objc
> ///
> /// Upon building the RoadGeometry, a Connection yields a Segment
> /// bearing multiple Lanes with offsets of the reference path. The
> ```

"offsets _from_ the reference path"

---

*[drake/automotive/maliput/multilane/builder.h, line 220 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbhqJOvwVultEM7GVa:-KvbhqJOvwVultEM7GVb:b-hzarsb) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L220)):*
> ```objc
>   /// Constructs a line-segment connection joining @p start to @p end.
>   ///
>   /// @param num_lanes The number of lanes. It must be grater than zero.
> ```

"greater" (META throughout the file.  At least it's consistent.  :) )

---

*[drake/automotive/maliput/multilane/builder.h, line 398 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbjKx5vC_IGIgUNcDW:-KvbjKx6Gti9LwafPAWN:b-pvkp3a) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L398)):*
> ```objc
>   /// derived from the composition of left and right shoulders, number of lanes
>   /// and spacing between Lanes. In addition, `r_spacing` will become the width
>   /// of all the Lanes, and lane bounds will be assigned over the centerline.
> ```

Hmm... reading this, if (finally?) occurs to me that the `r_spacing` parameter should just be called `lane_width` --- much more intuitive (for the user).  And then mention that centerlines are centered within the lanes, and that `lane_width` also becomes the spacing between centerlines.

---

*[drake/automotive/maliput/multilane/loader.cc, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc1GwE2Mkx2ultlx6h:-Kvc1GwF-Gyw8JIV_Rc2:boz6rcl) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/loader.cc#L158)):*
> ```cpp
> 
>   const api::RBounds lane_bounds(r_bounds(mmb["lane_bounds"]));
>   const api::RBounds driveable_bounds(r_bounds(mmb["lane_bounds"]));
> ```

This should be `mmb["driveable_bounds"]`.

It would be good to also assert that `lane_bounds` is symmetric (max == -min), since the new Builder can only handle the symmetric case. (The monolane Builder should not have allowed the asymmetric case anyway, and thankfully I don't think anyone ever used an asymmetric case.)

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc5IgLrmFcxDtATk-8:-Kvc5IgLrmFcxDtATk-9:b470t8q) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L35)):*
> ```cpp
> 
>   Endpoint start {{0., 0., -M_PI / 4.}, kLowFlatZ};
>   auto c0 = b.Connect("0", 1, 0., kLeftShoulder, kRightShoulder, start, 50.,
> ```

META this file: Constants for the common `1, 0,` parameters would help, too (e.g., `kOneLane, kZeroR0,`).

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 264 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc9wqla0n4oSL6IjiR:-Kvc9wqm-eRM77VJGp34:brkojcn) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L264)):*
> ```cpp
>   for (int i = 0; i < kNumLanes; i++) {
>     EXPECT_EQ(rg->junction(0)->segment(0)->lane(i)->id().string(),
>               std::string("l:c0_") + std::to_string(i));
> ```

It would be good to do a cursory check that, say, the starting point of each Lane ((s,r,h) = (0,0,0)) is in the right place (e.g., to test that the lateral-offset-per-lane-index logic is correct).

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 270 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc8z_hXCtdByIAoCBg:-Kvc8z_iMbWLp2O_oY9B:b-aa08hw) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L270)):*
> ```cpp
>     const api::BranchPoint* bp = rg->branch_point(i);
>     EXPECT_EQ(bp->id().string(), std::string("bp:") + std::to_string(i));
>     const api::Lane* lane = rg->junction(0)->segment(0)->lane(i / 2);
> ```

There is no requirement that the BranchPoints are constructed in the order tested here, so it's not clear what this chunk of code is actually testing.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 303 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcAJhiez2dV_GqVprF:-KvcAJhiez2dV_GqVprG:brkojcn) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L303)):*
> ```cpp
>   for (int i = 0; i < kNumLanes; i++) {
>     EXPECT_EQ(rg->junction(0)->segment(0)->lane(i)->id().string(),
>               std::string("l:c0_") + std::to_string(i));
> ```

It would be good to do a cursory check that, say, the starting point of each Lane ((s,r,h) = (0,0,0)) is in the right place (e.g., to test that the lateral-offset-per-lane-index logic is correct).

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 309 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc9erNh8bEFtaZ7zp_:-Kvc9erNh8bEFtaZ7zpa:b-aa08hw) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L309)):*
> ```cpp
>     const api::BranchPoint* bp = rg->branch_point(i);
>     EXPECT_EQ(bp->id().string(), std::string("bp:") + std::to_string(i));
>     const api::Lane* lane = rg->junction(0)->segment(0)->lane(i / 2);
> ```

There is no requirement that the BranchPoints are constructed in the order tested here, so it's not clear what this chunk of code is actually testing.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 335 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcChGHdGOlgADldXXH:-KvcChGIe1oNZyvEGAhK:bfvmawh) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L335)):*
> ```cpp
> // connections that later are used in code. 'x' characters are used to locate
> // BranchPoints. '0', '1' and '2' mean Lane indexes. Finally, '-', '|' and '\'
> // are used to draw Lane centerlines.
> ```

It would help a lot to also use letters `a - g` to indicate (roughly) where the named `endpoint_*` are.  (Use those letters instead of an `x` where appropriate.)

---

*[drake/automotive/maliput/multilane/test/multilane_lanes_test.cc, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbzsTqGET-vYEYFcCf:-KvbzsTqGET-vYEYFcCg:b-lop8bv) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_lanes_test.cc#L51)):*
> ```cpp
>   const double kHalfLaneWidth{5.};
>   const int kNumLanes{1};
>   const double kRSpacing{0.};
> ```

The only diff to this file in this PR is to add these two constants `kNumLanes` and `kRSpacing`.  Are they actually being used anywhere?

---

*[drake/automotive/maliput/multilane/test/multilane_road_geometry_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc4L0KrfZIxvdUhoue:-Kvc4L0KrfZIxvdUhouf:b-bqmyb6) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_road_geometry_test.cc#L50)):*
> ```cpp
>   const double kLength{50.};
>   const auto& lane0 =
>       rb->Connect("lane0", 1, 0., 0., 0., kRoadOrigin,
> ```

META in this file:  please define some constants to make the new parameters readable: `... kOneLane, kZeroR0, kNoShoulder, kNoShoulder, ...` or something like that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-Kvcgz6BRwJOnft9XdGo:b-een0iy)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcNqGQAaiegry2xYkr:-Kvd46ZhlaMcq6hW3ZqA:b-6q4s7p) (no related file):*
In my opinion there are two different functionalities. One comes from the point of view of the map creator and the other comes from the map parser / interpreter. Having loader tools that use the `Builder` class made me thought that the purpose of the `Builder` is to help building a `RoadGeometry` from a map, not to create it from scratch. Having said that, I think that a new approach should be taken (maybe adding more stuff is enough?) if the purpose of the `Builder` class is to be useful for both needs: help the parser to create a `RoadGeometry` from a map description and help the map designer to do the math.

> We need to support chaining connections from Lane to Lane that allows the reference curves to be disconnected -- so that the segments can twist around different axes in superelevation!

I think that if the Connection is capable of computing _start_ and _end_ `Endpoints` for each `Lane,` that will give enough information to create connections on a connected-lane basis. 

> I started trying to work out a nice representation that captures these constraints, but I haven't gotten it quite figured out yet. Perhaps it's best in a follow-up PR anyhow. Thoughts?

I agree. We can create an issue to discuss the needs and cases to cover. Although this revised interface may not suit all the needs I think it is enough to build a Multilane `RoadGeometry`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcNqGQAaiegry2xYkr:-KvgrIj3QTDAEpqoFtDn:brt6rd8) (no related file):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

In my opinion there are two different functionalities. One comes from the point of view of the map creator and the other comes from the map parser / interpreter. Having loader tools that use the `Builder` class made me thought that the purpose of the `Builder` is to help building a `RoadGeometry` from a map, not to create it from scratch. Having said that, I think that a new approach should be taken (maybe adding more stuff is enough?) if the purpose of the `Builder` class is to be useful for both needs: help the parser to create a `RoadGeometry` from a map description and help the map designer to do the math.

> We need to support chaining connections from Lane to Lane that allows the reference curves to be disconnected -- so that the segments can twist around different axes in superelevation!

I think that if the Connection is capable of computing _start_ and _end_ `Endpoints` for each `Lane,` that will give enough information to create connections on a connected-lane basis. 

> I started trying to work out a nice representation that captures these constraints, but I haven't gotten it quite figured out yet. Perhaps it's best in a follow-up PR anyhow. Thoughts?

I agree. We can create an issue to discuss the needs and cases to cover. Although this revised interface may not suit all the needs I think it is enough to build a Multilane `RoadGeometry`.


</blockquote></details>

Yes, the (original?) intent of the `Builder` class is to make it easier for people to programatically construct road networks --- by providing a language that expresses the common/simple ways people might think of hooking up the pieces together.  

I thought about it a bit more last night, and I think I have figured out the elements of a compact interface that captures the basic use cases.  I will transcribe this into an issue and then post the link here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcNqGQAaiegry2xYkr:-Kvh8Iu9Ruq5I8PEkQ_7:betequg) (no related file):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Yes, the (original?) intent of the `Builder` class is to make it easier for people to programatically construct road networks --- by providing a language that expresses the common/simple ways people might think of hooking up the pieces together.  

I thought about it a bit more last night, and I think I have figured out the elements of a compact interface that captures the basic use cases.  I will transcribe this into an issue and then post the link here.
</blockquote></details>

See #7194.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks for the review. PTAL.

---

Review status: 0 of 7 files reviewed at latest revision, 17 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvbn6nC_fEfOBtTzTPJ:-KvgsFCG8ynPtj9aAy_m:bgxjpzn) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.cc#L26)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This assertion already occurs within `HBounds()` itself.
</blockquote></details>

Done. However, if someone uses `HBounds` default constructor and accidentally uses `set_max()` and `set_min()` methods, it could be corrupted. Shouldn't it be also asserted on the setters of `HBounds`?

---

*[drake/automotive/maliput/multilane/builder.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbvRUBwvn6DdNHFiO3:-Kvlz69Wqmo_7cAqEKa4:b15cf25) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.cc#L124)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This use of the connection's/segment's `zpoint` here is incorrect.  It gives the `z` and `z'` values on the Segment's reference curve, but if superelevation is non-zero, then those values will be different on a Lane offset laterally.
</blockquote></details>

Done. I completely missed it. Thanks for pointing it out.

---

*[drake/automotive/maliput/multilane/builder.h, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbgmUW8xq8mg8iVX9Q:-KvgthTOg5alsRE4cGaE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L51)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Suggestion for clarity: 
```
  - Junctions: "j:" + Group::id(), or "j" + Connection::id() for an ungrouped Connection.
```
and then remove the `Note: when Connections are ungrouped...` comment below, too.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 202 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbhMpBhH3a4x8ua1Pf:-Kvgt8aXhvitGGAVs9X2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L202)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

"offsets _from_ the reference path"
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 220 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbhqJOvwVultEM7GVa:-KvgtVcdqxUZ0gG1Moxc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L220)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

"greater" (META throughout the file.  At least it's consistent.  :) )
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 398 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbjKx5vC_IGIgUNcDW:-KvgwAp0_Be6U5DcL_js:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.h#L398)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Hmm... reading this, if (finally?) occurs to me that the `r_spacing` parameter should just be called `lane_width` --- much more intuitive (for the user).  And then mention that centerlines are centered within the lanes, and that `lane_width` also becomes the spacing between centerlines.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/loader.cc, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc1GwE2Mkx2ultlx6h:-KvgxGGIwl-R9uxXE7qg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/loader.cc#L158)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This should be `mmb["driveable_bounds"]`.

It would be good to also assert that `lane_bounds` is symmetric (max == -min), since the new Builder can only handle the symmetric case. (The monolane Builder should not have allowed the asymmetric case anyway, and thankfully I don't think anyone ever used an asymmetric case.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc5IgLrmFcxDtATk-8:-Kvi1niyyVxeCpOvj757:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L35)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

META this file: Constants for the common `1, 0,` parameters would help, too (e.g., `kOneLane, kZeroR0,`).
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 264 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc9wqla0n4oSL6IjiR:-Kvi1qdcsRkQ-Nd0S8qY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L264)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

It would be good to do a cursory check that, say, the starting point of each Lane ((s,r,h) = (0,0,0)) is in the right place (e.g., to test that the lateral-offset-per-lane-index logic is correct).
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 270 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc8z_hXCtdByIAoCBg:-Kvi1w66VSQPFGl3Gyxi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L270)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

There is no requirement that the BranchPoints are constructed in the order tested here, so it's not clear what this chunk of code is actually testing.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 303 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcAJhiez2dV_GqVprF:-KvhXgLCvpvA2dqIqXXk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L303)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

It would be good to do a cursory check that, say, the starting point of each Lane ((s,r,h) = (0,0,0)) is in the right place (e.g., to test that the lateral-offset-per-lane-index logic is correct).
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 309 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc9erNh8bEFtaZ7zp_:-KvhxHPG-aBSAVnWMhsO:bdn8th2) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L309)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

There is no requirement that the BranchPoints are constructed in the order tested here, so it's not clear what this chunk of code is actually testing.
</blockquote></details>

Done. Moved inside the for to be part of the lane checks.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 335 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcChGHdGOlgADldXXH:-KvhyBPlRzGLnjJtIL3-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L335)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

It would help a lot to also use letters `a - g` to indicate (roughly) where the named `endpoint_*` are.  (Use those letters instead of an `x` where appropriate.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_road_geometry_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvc4L0KrfZIxvdUhoue:-Kvi0CjxbCICtRq__5U_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_road_geometry_test.cc#L50)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

META in this file:  please define some constants to make the new parameters readable: `... kOneLane, kZeroR0, kNoShoulder, kNoShoulder, ...` or something like that.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_lanes_test.cc, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvbzsTqGET-vYEYFcCf:-Kvi-FpWKVgqzuhzeTuF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/test/multilane_lanes_test.cc#L51)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

The only diff to this file in this PR is to add these two constants `kNumLanes` and `kRSpacing`.  Are they actually being used anywhere?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KvmqKaYlg2akB1yA-zQ:b-k49e1v)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Review status: 0 of 7 files reviewed at latest revision, 17 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KvcNP4iWnUYjAXUQCD4:-KvmqTJDYYbT-xn5527a:b-896fix) (no related file):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

It would help to mention in the commit message that the YAML format changes will come in a subsequent PR.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  (modulo one lane-offset-related bug)

---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-Kvbn6nC_fEfOBtTzTPJ:-Kw7XZ4HienApO2BtTkX:bp2jz3i) ([raw file](https://github.com/robotlocomotion/drake/blob/3e82bf5b38fd61875bf79422e01c93464b736ffe/drake/automotive/maliput/multilane/builder.cc#L26)):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

Done. However, if someone uses `HBounds` default constructor and accidentally uses `set_max()` and `set_min()` methods, it could be corrupted. Shouldn't it be also asserted on the setters of `HBounds`?
</blockquote></details>

Good catch!  It looks like `RBounds` has the same problem.  I filed #7231.  (Feel free to self-assign if it is really bothering you. :)

---

*[drake/automotive/maliput/multilane/builder.cc, line 144 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBBdsqDNDC9-v9LEHU:-KwBBdsrm6N7WSBT9y43:b-dlp515) ([raw file](https://github.com/robotlocomotion/drake/blob/dfb746ecce286e0669081ac9774cd156d8a772e2/drake/automotive/maliput/multilane/builder.cc#L144)):*
> ```cpp
>   const Vector3<double> w_prime =
>       road_curve->W_prime_of_prh(p, 0., 0., road_curve->Rabg_of_p(p),
>                                  road_curve->elevation().f_dot_p(p));
> ```

The `r` argument here should be the lane's lateral offset, otherwise it is calculating W' on the ref. curve.  (`position` and `rotation` are ok because they come from asking `lane` itself.)

---

*[drake/automotive/maliput/multilane/segment.h, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBDIqwKrXeLJR4LRPC:-KwBDIqwKrXeLJR4LRPD:b-fm25m1) ([raw file](https://github.com/robotlocomotion/drake/blob/dfb746ecce286e0669081ac9774cd156d8a772e2/drake/automotive/maliput/multilane/segment.h#L74)):*
> ```objc
>   Lane* NewLane(api::LaneId id, double r0, const api::RBounds& lane_bounds);
> 
>   const RoadCurve* road_curve() const { return road_curve_.get(); }
> ```

BTW:  if the only reason for this accessor is for `ComputeEndpointForLane()` in `builder.cc`, one could get around that by passing the `RoadCurve` pointer directly into `ComputeEndpointForLane()` (and thus avoiding the dynamic cast in there, too).  It's a possibility....

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwBIpcC0CV-Po5weSwZ:b6ytt9s)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks for the reviews.

---

Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 144 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBBdsqDNDC9-v9LEHU:-KwBQzS9OiOeIgHE_sip:beg7d7b) ([raw file](https://github.com/robotlocomotion/drake/blob/dfb746ecce286e0669081ac9774cd156d8a772e2/drake/automotive/maliput/multilane/builder.cc#L144)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

The `r` argument here should be the lane's lateral offset, otherwise it is calculating W' on the ref. curve.  (`position` and `rotation` are ok because they come from asking `lane` itself.)
</blockquote></details>

Done. Yes, I missed it.

---

*[drake/automotive/maliput/multilane/segment.h, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBDIqwKrXeLJR4LRPC:-KwBQyUTNbYBDu7RqvdC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dfb746ecce286e0669081ac9774cd156d8a772e2/drake/automotive/maliput/multilane/segment.h#L74)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

BTW:  if the only reason for this accessor is for `ComputeEndpointForLane()` in `builder.cc`, one could get around that by passing the `RoadCurve` pointer directly into `ComputeEndpointForLane()` (and thus avoiding the dynamic cast in there, too).  It's a possibility....
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwBWRB4fWeQxc_bIZmx:b-hlc8b9)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
+@seancurtis-tri for platform review please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwBZW7urIvc8qvMFnnY:b-tnzbcp)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> platform with a few cosmetic issues.

---

Reviewed 6 of 8 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 122 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC6uxJdn7E--TLYRa_:-KwC6uxJdn7E--TLYRaa:buap7ie) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.cc#L122)):*
> ```cpp
> 
> // Computes the location and heading of a `lane` at given `end` creating an
> // Endpoint with that information. `zpoint` is passed here as is already
> ```

I'm not sure this sentence makes sense -- as a `const EndpointZ&` it should be apparent that this method won't change it. Perhaps just cut it?

---

*[drake/automotive/maliput/multilane/builder.cc, line 125 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC6LhjYZI7chPwr4gr:-KwC6LhjYZI7chPwr4gs:b-6wxs01) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.cc#L125)):*
> ```cpp
> // computed. `road_curve` is used to compute z derivative with respect to p at
> // the start or end of the `lane` respectively.
> Endpoint ComputeEndpointForLane(const RoadCurve* road_curve, const Lane* lane,
> ```

`const RoadCurve& road_curve`

---

*[drake/automotive/maliput/multilane/builder.h, line 272 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC4S2EJ_lEanorsGvy:-KwC4S2EJ_lEanorsGvz:b7duil4) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.h#L272)):*
> ```objc
>     DRAKE_DEMAND(num_lanes_ > 0);
>     DRAKE_DEMAND(left_shoulder_ >= 0);
>     DRAKE_DEMAND(right_shoulder_ >= 0);
> ```

Non-negative radius not enforced. `DRAKE_DEMAND(radius_ >= 0);`

---

*[drake/automotive/maliput/multilane/builder.h, line 420 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC5O3IvtRXfw8YKcFw:-KwC5O3IvtRXfw8YKcFx:b7ib4l8) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.h#L420)):*
> ```objc
>   /// characteristics at the end-point.
>   /// `r0` is the distance from the reference curve to the first Lane
>   /// centerline. `left_shoulder` and `right_shoulder` are extra lateral
> ```

FYI Meta: this document is now quite erratic in using `@p` and backticks.

---

*[drake/automotive/maliput/multilane/segment.h, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBn05pBSA0TTBbclxU:-KwBn05pBSA0TTBbclxV:b-9lg0e1) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/segment.h#L74)):*
> ```objc
>   Lane* NewLane(api::LaneId id, double r0, const api::RBounds& lane_bounds);
> 
>   const RoadCurve* road_curve() const { return road_curve_.get(); }
> ```

Why a pointer?  This is only used as an argument to `ComputeEndpointForLane` (which breaks the style guide in passing a const pointer) and doesn't confirm the pointer is non-null. This should return a const ref and that should propagate through.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 165 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC7vI4wY7AdihunTqm:-KwC7vI4wY7AdihunTqn:b-p3x0hq) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L165)):*
> ```cpp
>   //
>   // Make up some combination of default routes:
>   b.SetDefaultBranch(left1, 0, api::LaneEnd::kStart, left1, 0,
> ```

BTW The use of `kOneLane` and `kZeroR0` make the magical `0`s in this code stand out.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 295 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC8mgvEKUOETW2M2vJ:-KwC8mgvEKUOETW2M2vK:bpmjfr2) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L295)):*
> ```cpp
>     // Checks lane start geo position to verify that spacing is correctly
>     // applied.
>     const Vector3<double> lane_start_geo = start_reference_curve +
> ```

BTW Got some clang formatting issues going on here. I'd just hit  all of the new methods with clang format so I can save calling them out. (There are about five or six.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwCBkOHy7D7-5TmuKRG:b-hzoswq)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/automotive/maliput/multilane/builder.cc, line 287 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwE4hiTG5GjZ1kFl9GK:-KwE4hiTG5GjZ1kFl9GL:b45s7m) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.cc#L287)):*
> ```cpp
>         segment->road_curve(), lane, api::LaneEnd::kStart, conn->start().z());
>     const Endpoint finish_endpoint = ComputeEndpointForLane(
>         segment->road_curve(), lane, api::LaneEnd::kFinish, conn->end().z());
> ```

Instead of calling `segment->road_curve()`, just capture/alias `road_curve`'s value before the `std::move()` into the new Segment.

(See comment in `segment.h` below.)

---

*[drake/automotive/maliput/multilane/segment.h, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBn05pBSA0TTBbclxU:-KwE3i3nAKsCVj6e6Wv8:blrft2k) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/segment.h#L74)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Why a pointer?  This is only used as an argument to `ComputeEndpointForLane` (which breaks the style guide in passing a const pointer) and doesn't confirm the pointer is non-null. This should return a const ref and that should propagate through.
</blockquote></details>

In fact... this method is not really needed anywhere.  `ComputeEndpointForLane` is only called within `BuildConnection`, and `BuildConnection` already has access to the `RoadCurve` because it constructed it.  (See clarifying comment up in `builder.cc`.)

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 295 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC8mgvEKUOETW2M2vJ:-KwE5Mtok-fP-ggsiIeZ:bz8jbwr) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L295)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Got some clang formatting issues going on here. I'd just hit  all of the new methods with clang format so I can save calling them out. (There are about five or six.)
</blockquote></details>

(Emphasis on "the new methods" --- nothing more excruciating than trying to read a diff that has random unrelated clang-format-bombs scattered throughout.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 295 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC8mgvEKUOETW2M2vJ:-KwFocTqqsBMWw6uYiTE:b-8av9qx) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L295)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

(Emphasis on "the new methods" --- nothing more excruciating than trying to read a diff that has random unrelated clang-format-bombs scattered throughout.)
</blockquote></details>

Amen!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Couple typos...

---

Reviewed 6 of 7 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/automotive/maliput/multilane/builder.h, line 272 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC4S2EJ_lEanorsGvy:-KwGOu2w4VZSK2wQKIB2:b6x15rx) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.h#L272)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Non-negative radius not enforced. `DRAKE_DEMAND(radius_ >= 0);`
</blockquote></details>

The documentation says "non-negative", this test says "positive". Needs to be reconciled. (Personally, I think positive makes more sense.)

---

*[drake/automotive/maliput/multilane/builder.h, line 222 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwGO_gpiOl7B5cqqUNx:-KwGO_gpiOl7B5cqqUNy:bfl6j8r) ([raw file](https://github.com/robotlocomotion/drake/blob/72915a9aad41706e983ae192c2995f568e01329f/drake/automotive/maliput/multilane/builder.h#L222)):*
> ```objc
>   /// curve.
>   ///
>   /// `left_shoulder` and `left_shoulder` are extra spaces added to the right
> ```

One of these should probably be `right_shoulder`.

---

*[drake/automotive/maliput/multilane/builder.h, line 253 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwGOn0voXjufU9TN-Ck:-KwGOn0voXjufU9TN-Cl:b-hrf2sd) ([raw file](https://github.com/robotlocomotion/drake/blob/72915a9aad41706e983ae192c2995f568e01329f/drake/automotive/maliput/multilane/builder.h#L253)):*
> ```objc
>   /// curve.
>   ///
>   /// `left_shoulder` and `left_shoulder` are extra spaces added to the right
> ```

Also lost `right_shoulder` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwGPNTDEbKOLQIGOFfO:b-ghvv9i)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I've addressed the comments. PTAL.

---

Review status: 5 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/automotive/maliput/multilane/builder.cc, line 122 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC6uxJdn7E--TLYRa_:-KwFr3rl0li3y6nvFqR0:bs7wumx) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.cc#L122)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'm not sure this sentence makes sense -- as a `const EndpointZ&` it should be apparent that this method won't change it. Perhaps just cut it?
</blockquote></details>

Done. 

---

*[drake/automotive/maliput/multilane/builder.cc, line 125 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC6LhjYZI7chPwr4gr:-KwFqpMkdZGmuFVtew8e:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.cc#L125)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

`const RoadCurve& road_curve`
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.cc, line 287 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwE4hiTG5GjZ1kFl9GK:-KwFsU8I1xjRuJ5XvDFc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.cc#L287)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Instead of calling `segment->road_curve()`, just capture/alias `road_curve`'s value before the `std::move()` into the new Segment.

(See comment in `segment.h` below.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 272 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC4S2EJ_lEanorsGvy:-KwFswpYzIJmSnMfGVQO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.h#L272)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

The documentation says "non-negative", this test says "positive". Needs to be reconciled. (Personally, I think positive makes more sense.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 420 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC5O3IvtRXfw8YKcFw:-KwG-ERG-xupkezvwzGM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/builder.h#L420)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Meta: this document is now quite erratic in using `@p` and backticks.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 222 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwGO_gpiOl7B5cqqUNx:-KwGQrDuYMrB8z5nfb7q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/72915a9aad41706e983ae192c2995f568e01329f/drake/automotive/maliput/multilane/builder.h#L222)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

One of these should probably be `right_shoulder`.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/builder.h, line 253 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwGOn0voXjufU9TN-Ck:-KwGQUqQLiDW1qnaboD7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/72915a9aad41706e983ae192c2995f568e01329f/drake/automotive/maliput/multilane/builder.h#L253)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Also lost `right_shoulder` here.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 165 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC7vI4wY7AdihunTqm:-KwFvi3JVQSPyOgTAJgb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L165)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The use of `kOneLane` and `kZeroR0` make the magical `0`s in this code stand out.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_builder_test.cc, line 295 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwC8mgvEKUOETW2M2vJ:-KwFyu2HAF6sBAZmgy9f:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/test/multilane_builder_test.cc#L295)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Amen!
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/segment.h, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwBn05pBSA0TTBbclxU:-KwFuwEk7ZPvSjAJRX2l:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09e7ff29e26439b749c0c8ae44f7bc24731ed481/drake/automotive/maliput/multilane/segment.h#L74)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

In fact... this method is not really needed anywhere.  `ComputeEndpointForLane` is only called within `BuildConnection`, and `BuildConnection` already has access to the `RoadCurve` because it constructed it.  (See clarifying comment up in `builder.cc`.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwGRs9cZ4zYj_2uYZfB:b-3sgecx)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Reviewed 1 of 7 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Two last typos have crept in.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[drake/automotive/maliput/multilane/builder.h, line 224 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwGd8gUTaOfXqAdK9Ge:-KwGd8gUTaOfXqAdK9Gf:b-ebm6hr) ([raw file](https://github.com/robotlocomotion/drake/blob/0f1e75d6b21150b7730b00da3c2edca44bbfdf59/drake/automotive/maliput/multilane/builder.h#L224)):*
> ```objc
>   /// `left_shoulder` and `right_shoulder` are extra spaces added to the right
>   /// and left side of the first and last lanes of the Segment. They will be
>   /// added to Segment's bounds and must be must be greater or equal to zero.
> ```

typo: "must be must be"

---

*[drake/automotive/maliput/multilane/builder.h, line 255 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-KwGdD5JsWvmRSKaShDo:-KwGdD5JsWvmRSKaShDp:b-3s60pz) ([raw file](https://github.com/robotlocomotion/drake/blob/0f1e75d6b21150b7730b00da3c2edca44bbfdf59/drake/automotive/maliput/multilane/builder.h#L255)):*
> ```objc
>   /// `left_shoulder` and `right_shoulder` are extra spaces added to the right
>   /// and left side of the first and last lanes of the Segment. They will be
>   /// added to Segment's bounds and must be must be greater or equal to zero.
> ```

typo "must be must be"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwGdNCTn-mI0OdZZKif:b-g7holv)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
No worries.  :)

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7171#-:-KwGj4CWW379fIaBLkWh:b-lj9kug)*
<!-- Sent from Reviewable.io -->
